### PR TITLE
Add filter locator and deprecate filter collection

### PR DESCRIPTION
**Pull request by meyerbaptiste**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | N/A
| License       | MIT
| Doc PR        | N/A

First step before #1082 and #1083, we have to refactor the `FilterCollection` to avoid circular references with new filters.
Our solution is to use the PSR11 and a service locator. Thanks to http://symfony.com/blog/new-in-symfony-3-3-service-locators.

### CODE REVIEW

File: src/Api/FilterCollectionFactory.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+     */
+    public function createFilterCollectionFromLocator(ContainerInterface $filterLocator): FilterCollection
+    {
+        $filters = [];
+
+        foreach ($this->filters as $filter) {
+            try {
+                $filters[$filter] = $filterLocator->get($filter);
+            } catch (\Throwable $e) {
+            }
```
Comment (soyuka) [inline src/Api/FilterCollectionFactory.php]: Shouldn't we use `if ($filterLocator->has($filter)) {}` instead of catching?

File: src/Bridge/Doctrine/Orm/Extension/FilterExtension.php [hunk 2]
```diff
@@ -27,13 +27,15 @@
  */
 final class FilterExtension implements QueryCollectionExtensionInterface
 {
+    use FilterLocatorTrait;
+
     private $resourceMetadataFactory;
-    private $filters;
 
-    public function __construct(ResourceMetadataFactoryInterface $resourceMetadataFactory, FilterCollection $filters)
+    public function __construct(ResourceMetadataFactoryInterface $resourceMetadataFactory, $filterLocator)
```
Comment (soyuka) [inline src/Bridge/Doctrine/Orm/Extension/FilterExtension.php]: There is no such typing as `ContainerInterface | FilterCollection` in php right? Maybe we could document this new argument though.

File: src/Metadata/Extractor/YamlExtractor.php [hunk 3]
```diff
@@ -32,7 +32,7 @@
     protected function extractPath(string $path)
     {
         try {
-            $resourcesYaml = Yaml::parse(file_get_contents($path));
+            $resourcesYaml = Yaml::parse(file_get_contents($path), Yaml::PARSE_KEYS_AS_STRINGS);
```
Comment (soyuka) [inline src/Metadata/Extractor/YamlExtractor.php]: Hmm why this change? Did you hit a bug? This might need to be backported to 2.0 if it's the case.

File: src/Bridge/Doctrine/Orm/Extension/FilterExtension.php (hunk 2 above)
Comment (soyuka) [inline src/Bridge/Doctrine/Orm/Extension/FilterExtension.php]: same goes in the other classes with the same change

Comment (meyerbaptiste) [inline src/Bridge/Doctrine/Orm/Extension/FilterExtension.php]: This works only for try/catch with PHP 7.1. I will add some PHPDoc.

File: src/Metadata/Extractor/YamlExtractor.php (hunk 3 above)
Comment (meyerbaptiste) [inline src/Metadata/Extractor/YamlExtractor.php]: I have deprecation warnings from the component without this change and tests fail...

Comment (soyuka) [inline src/Metadata/Extractor/YamlExtractor.php]: oh okay this is because symfony 3.3 then alright! Thanks for the information.

Comment (meyerbaptiste) [inline src/Metadata/Extractor/YamlExtractor.php]: https://github.com/symfony/symfony/pull/21774

File: src/Bridge/Doctrine/Orm/Extension/FilterExtension.php (hunk 2 above)
Comment (meyerbaptiste) [inline src/Bridge/Doctrine/Orm/Extension/FilterExtension.php]: I just made the change. Is it good for you like this?

Comment (soyuka) [inline src/Bridge/Doctrine/Orm/Extension/FilterExtension.php]: :+1: I hope that phpstan is fine with it and yeah this is better for future developers IMO. Thanks!

**Comment by meyerbaptiste** [comment]
Tests fail because the master branch fails too.

File: src/Bridge/Symfony/Bundle/Resources/config/filter.xml [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+           xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
+
+    <services>
+        <service id="api_platform.filter_locator" class="Symfony\Component\DependencyInjection\ServiceLocator" public="false">
+            <tag name="container.service_locator" />
+        </service>
+
+        <service id="api_platform.filter_collection_factory" class="ApiPlatform\Core\Api\FilterCollectionFactory" public="false" />
+
+        <service id="api_platform.filters" class="ApiPlatform\Core\Api\FilterCollection" public="false">
```
Comment (meyerbaptiste) [inline src/Bridge/Symfony/Bundle/Resources/config/filter.xml]: Oh, nice, I missed this feature! Thanks!

**Comment by soyuka** [comment]
> Tests fail because the master branch fails too.

Indeed, the `Accept` header isn't changing the format. [This](https://github.com/api-platform/core/blob/master/features/doctrine/search_filter.feature#L10) returns JSON-LD. It's weird, everything was green, then I just did `composer update` and it now fails.

Investigating but maybe @dunglas has a clue on this?

File: composer.json [hunk 5]
```diff
@@ -23,7 +23,7 @@
         "symfony/property-access": "^2.7 || ^3.0",
         "symfony/property-info": "^3.3@beta",
         "symfony/serializer": "^3.3@beta",
-        "willdurand/negotiation": "^2.0.3"
+        "willdurand/negotiation": "~2.2.0"
```
Comment (soyuka) [inline composer.json]: nice one maybe we can patch this in a separated pr?

**Comment by meyerbaptiste** [comment]
Ok, I found it. The problem comes from the new release of `willdurand/negotiation` (2.3.0) and this commit: https://github.com/willdurand/Negotiation/pull/92.

File: composer.json (hunk 5 above)
Comment (dunglas) [inline composer.json]: Should be `~2.2.3` right?

Comment (meyerbaptiste) [inline composer.json]: What about `>=2.0.3 <2.3.0`?

Comment (teohhanhui) [inline composer.json]: `^2.0.3 <2.3`

**Comment by meyerbaptiste** [comment]
Does anyone have an idea to fix/avoid that?
```
Generating code coverage report in PHP format ... done
Remaining deprecation notices (1)
The ApiPlatform\Core\Api\FilterCollection class is deprecated since version 2.1 and will be removed in 3.0. Implement the Psr\Container\ContainerInterface interface as service locator instead: 1x
    1x in CodeCoverage::start from SebastianBergmann\CodeCoverage
```

**Comment by chalasr** [comment]
@meyerbaptiste CodeCoverage loads classes which causes the FilterCollection deprecation to be triggered. What you need is to move the deprecation to the constructor:

```diff
diff --git a/src/Api/FilterCollection.php b/src/Api/FilterCollection.php
index 32b1157..76af317 100644
--- a/src/Api/FilterCollection.php
+++ b/src/Api/FilterCollection.php
@@ -15,8 +15,6 @@ namespace ApiPlatform\Core\Api;
 
 use Psr\Container\ContainerInterface;
 
-@trigger_error(sprintf('The %s class is deprecated since version 2.1 and will be removed in 3.0. Implement the %s interface as service locator instead.', FilterCollection::class, ContainerInterface::class), E_USER_DEPRECATED);
-
 /**
  * A list of filters.
  *
@@ -26,4 +24,10 @@ use Psr\Container\ContainerInterface;
  */
 final class FilterCollection extends \ArrayObject
 {
+    public function __construct($input = [], $flags = 0, $iterator_class = \ArrayIterator::class)
+    {
+        @trigger_error(sprintf('The %s class is deprecated since version 2.1 and will be removed in 3.0. Implement the %s interface as service locator instead.', self::class, ContainerInterface::class), E_USER_DEPRECATED);
+
+        parent::__construct($input, $flags, $iterator_class);
+    }
 }

```

**Comment by soyuka** [comment]
Also you can mark specific tests as `@group legacy`  and use `@expectedDeprecation`

**Comment by meyerbaptiste** [comment]
It's what I made @soyuka but the problem here comes from coverage and not tests themselves.
But I think I will follow the @chalasr's suggestion and move my `trigger_error()` inside my controller to avoid the deprecation notice during the coverage.

**Comment by chalasr** [comment]
@soyuka comment made me think that maybe moving the `@group legacy`+`@expectedDeprecation` from the test methods to the top of the test class would solve it, unfortunately it does not. Either you have to trigger deprecations from constructors in order to don't have it triggered before construction, or the whole class should be excluded from the coverage, which I think is not a good idea as the class is still usable and well tested.

File: src/Api/IdentifiersExtractor.php [hunk 6]
```diff
@@ -76,6 +76,7 @@ public function getIdentifiersFromItem($item): array
                 throw new RuntimeException(sprintf('No identifier found in "%s" through relation "%s" of "%s" used as identifier', $relatedResourceClass, $propertyName, $resourceClass));
             }
         }
+
```
Comment (soyuka) [inline src/Api/IdentifiersExtractor.php]: Was fixed by https://github.com/api-platform/core/commit/7dfdcb1715210f12160919d92975e15d8da7e81f

File: features/main/crud.feature [hunk 7]
```diff
@@ -123,73 +123,67 @@ Feature: Create-Retrieve-Update-Delete
...
         "@type": "hydra:IriTemplate",
-        "hydra:template": "/dummies{?id,id[],name,alias,description,relatedDummy.name,relatedDummy.name[],relatedDummies,relatedDummies[],dummy,relatedDummies.name,order[id],order[name],order[relatedDummy.symfony],dummyDate[before],dummyDate[after],relatedDummy.dummyDate[before],relatedDummy.dummyDate[after],dummyFloat[between],dummyFloat[gt],dummyFloat[gte],dummyFloat[lt],dummyFloat[lte],dummyPrice[between],dummyPrice[gt],dummyPrice[gte],dummyPrice[lt],dummyPrice[lte],dummyBoolean,dummyFloat,dummyPrice,description[exists],relatedDummy.name[exists],dummyBoolean[exists]}",
+        "hydra:template": "\/dummies{?dummyBoolean,dummyDate[before],dummyDate[after],relatedDummy.dummyDate[before],relatedDummy.dummyDate[after],description[exists],relatedDummy.name[exists],dummyBoolean[exists],dummyFloat,dummyPrice,order[id],order[name],order[relatedDummy.symfony],dummyFloat[between],dummyFloat[gt],dummyFloat[gte],dummyFloat[lt],dummyFloat[lte],dummyPrice[between],dummyPrice[gt],dummyPrice[gte],dummyPrice[lt],dummyPrice[lte],id,id[],name,alias,description,relatedDummy.name,relatedDummy.name[],relatedDummies,relatedDummies[],dummy,relatedDummies.name}",
```
Comment (soyuka) [inline features/main/crud.feature]: Not sure what went wrong here? If this is due to random sorting maybe we should sort everything to avoid such issues in the feature?

Comment (dunglas) [inline features/main/crud.feature]: (we discussed this IRL with @meyerbaptiste).

Before: filters were in the same order than declared in `services.yml`
Now: filters are in the order the configuration in the resource.

It's not a BC break (the JSON parser should not depend of the order).
I prefer to not sort the filters, it allows to detect when we change the output, it's a desirable behavior.

File: src/Api/FilterCollection.php [hunk 8]
```diff
@@ -13,11 +13,21 @@
...
  *
  * @author Kévin Dunglas <dunglas@gmail.com>
+ *
+ * @deprecated since version 2.1, to be removed in 3.0. Use a service locator {@see \Psr\Container\ContainerInterface}.
  */
 final class FilterCollection extends \ArrayObject
 {
+    public function __construct($input = [], $flags = 0, $iterator_class = 'ArrayIterator')
+    {
+        @trigger_error(sprintf('The %s class is deprecated since version 2.1 and will be removed in 3.0. Implement the %s interface as service locator instead.', self::class, ContainerInterface::class), E_USER_DEPRECATED);
```
Comment (dunglas) [inline src/Api/FilterCollection.php]: I would say `Provide and implementation of "%s" instead`.

File: src/Api/FilterCollectionFactory.php [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+ *
+ * @author Baptiste Meyer <baptiste.meyer@gmail.com>
+ *
+ * @internal
+ */
+class FilterCollectionFactory
+{
+    private $filtersIds;
+
+    public function __construct(string ...$filtersIds)
```
Comment (dunglas) [inline src/Api/FilterCollectionFactory.php]: Why not using an array?

Comment (meyerbaptiste) [inline src/Api/FilterCollectionFactory.php]: Just a hack to have a typed array...

Comment (dunglas) [inline src/Api/FilterCollectionFactory.php]: I prefer to use a raw array and use `string[]` in the PHPDoc, PHPStan will be able to check types using the PHPDoc.

Comment (meyerbaptiste) [inline src/Api/FilterCollectionFactory.php]: Bad day for variadic functions 😭

Comment (theofidry) [inline src/Api/FilterCollectionFactory.php]: if you really want a check you could also force it _after_ in the code:

```php
$filterIds = (function (string ...$filterIds) { return $filterIds; })(...$filterIds);
```

Or use a lib like https://github.com/beberlei/assert

Comment (dunglas) [inline src/Api/FilterCollectionFactory.php]: We don't have to enforce this. The PHPDoc is the contract and it's an `@internal` class.

Comment (theofidry) [inline src/Api/FilterCollectionFactory.php]: I'm just giving alternatives to enforce it, not saying you have to enforce it :P

**Comment by meyerbaptiste** [comment]
All comments are fixed.

**Comment by dunglas** [comment]
Thanks @meyerbaptiste

**Comment by teohhanhui** [comment]
Unfortunately, this must be reverted because bumping the required versions in composer.json is a major BC break. It means we suddenly drop support for a lot of people... :disappointed:

**Comment by dunglas** [comment]
@teohhanhui it's not, its allowed by semver (and we do this everytime in Symfony). Note that we don't force to upgrade to new major versions, just to the last minor.

**Comment by teohhanhui** [comment]
But if we're thinking in semver, 2.1 is not a major version. It's a minor version.

**Comment by dunglas** [comment]
Semver is about code API. Not about dependencies. Every major project (including Symfony, Doctrine...) do this and this is totally acceptable. It's also the only way to no being stuck with very old and versions.
It's not a big deal for users to upgrade to a new minor version of a project. If they use the dependency themselves, it must be explicitly defined in their `composer.json` to prevent problems.

**Comment by teohhanhui** [comment]
I don't know if we'll ever reach agreement about the technicality of whether bumping dependencies is a BC break, but in the *spirit* of semver, it'd be best if we can tag a major release when we drop support for an older *minor* version of Symfony.

In any case, [stability flags](https://getcomposer.org/doc/04-schema.md#package-links) only work at the project root level. So when we use `@beta` in `require`, that does nothing and will result in unfulfillable dependencies (because Symfony 3.3 is not released yet).

**Comment by dunglas** [comment]
You're right about stability flags, but it's just a temporary solution. We'll not release 2.1 before the release of Symfony 3.3.

**Comment by teohhanhui** [comment]
Looks like you're right about semver:

http://semver.org/#what-should-i-do-if-i-update-my-own-dependencies-without-changing-the-public-api

:smile:
